docs: Improve code blocks

edibotopic opened the pull request:
This PR improves the consistency and functionality of code blocks in the docs.

The main changes are:

- Addition of a CSS selector that allows copy buttons to be deactivated in code-blocks. This is especially useful for code blocks that contain output.

A code block with the regular behaviour is written:

\```{code-block) text
PS C:\Users\myuser> echo $env:USERPROFILE
\```

Result (note the button on cursor hover): 

![codeblock-regular-behaviour](https://github.com/ubuntu/WSL/assets/66971213/6eed284a-c5ee-4bc2-8985-bb59cabf0b40)

If we do not want the copy button, we write:

\```{code-block) text
:class: no-copy
C:\Users\myuser
\```

Result (note the lack of button on cursor hover):

![output-no-copybtn](https://github.com/ubuntu/WSL/assets/66971213/ae4cffe7-16a0-4e9a-9bde-aca0e240475a)

- Addition of regex-based solution to prevent copy/pasting of command prompts. This ensures that the prompt and any text preceding the prompt is not copied when a user clicks the copy button. This is done automatically for any code block marked with the {code-block} directive.

The example below shows two lines with complex prompts and a symbol sometimes used as a prompt ($) in the commands (top). After the copy button is clicked, only the relevant information is pasted into the text editor (bottom). This means that users can more easily and quickly copy commands from the docs:

![prompt-paste-nocopy](https://github.com/ubuntu/WSL/assets/66971213/e0ccfcd5-dd85-4107-ac0c-d32f7d15ab72)

Other minor changes:

- Explicit use of code block directive in markup.
- Replacement of blockquotes and other unconventional methods of code markup.
- Where possible, separation of input and output, using text to explain which is which.
- Prefer no syntax highlighting for terminal commands.
- Delete any image that duplicates content in code blocks.
- Some minor fixes to style and wording.

UDENG-3441

CarlosNihelton reviewed (commented):
Very good job here. More than doing the changes the challenge seems to be reading throughout this documentation  set and figuring out where to apply the changes.

I have two minor points for your consideration.

edibotopic commented on docs/custom_conf.py:
```diff
@@ -224,3 +224,10 @@
+copybutton_selector = "div:not(.no-copy) > div.highlight > pre"
+
+# Define prompts to be excluded from copying when a copy button is used 
+copybutton_prompt_text = r"^.*?[\$>#]\s+"
```
Interesting edge case -- well spotted. How common are directory names like this? I wish people wouldn't name directories this way in the first place but we can't win that battle! I didn't encounter these cases in our docs afaik so didn't consider this. I tested your greedy alternative on a set of examples and it seems to work as before but solves this particular issue. I would be happy to go with your suggestion unless you have identified other pitfalls.

Ideally the person writing the documentation should not have to think about the regex in the background; however, I think it is good practice to test how your copyable material is actually copied/pasted. So perhaps in the docs contribution guidelines there could be a brief explanation as to why we use prompts in our docs (currently), how we use regex to solve the copy/paste issue and what to do if an edge-case is encountered.

I still need to explore in the future if there's a better way to communicate PS versus bash without heavy use of prompts. I don't think including the path in the prompt is always necessary but that's something for another day.

CarlosNihelton replied:
> How common are directory names like this? 

I doubt it's a common pattern for directory names :)

jibel replied:
If you remove the non-greedy marker then in `~$ echo "Hello" > file.txt`, for instance (redirections are a common operation), the prompt would be `~$ echo "Hello" > ` instead of just `~$`. It would be obviously wrong.

CarlosNihelton replied:
Very good point.

edibotopic replied:
I need to find time for more tests tomorrow but the lazy option worked for the types of prompts and commands that exist in the docs. When I tested the greedy alternative I should have considered a broader set of cases.

Current thinking is: lazy option with some wording in contributing.md about conventions for prompts and testing expected copy behavior.

edibotopic commented on docs/tutorials/cloud-init.md:
```diff
@@ -6,18 +6,18 @@ Ubuntu WSL users can now leverage it to perform an automatic setup to get a work
 > See more:  [cloud-init official documentation](https://cloudinit.readthedocs.io/en/latest/index.html).
 
-The latest release of Ubuntu (Noble Numbat 24.04 LTS) comes with cloud-init already preinstalled, so you'll need that specific application to follow this tutorial. Ubuntu 24.04 LTS can be installed from [this link to the Microsoft Store](https://www.microsoft.com/store/productId/9NZ3KLHXDJP5?ocid=pdpshare). A previous version of this tutorial used Ubuntu-Preview, because that comes with the lates ...
+The latest release of Ubuntu (Noble Numbat 24.04 LTS) comes with cloud-init already preinstalled, so you'll need that specific application to follow this tutorial. Ubuntu 24.04 LTS can be installed from [this link to the Microsoft Store](https://www.microsoft.com/store/productId/9NZ3KLHXDJP5?ocid=pdpshare). A previous version of this tutorial used Ubuntu Preview, because that comes with the lates ...
```
Yes -- agreed! In this case it should probably be "Ubuntu (Preview)", as it also agrees with the form used earlier in the paragraph. Thanks for this.

The naming convention for product releases versus instances is something for me to note down. Our style guide has some specific guidance on naming distributions. So the example above should be Ubuntu 24.04 LTS (Noble Numbat); however, it doesn't offer guidance on something like WSL instances of Ubuntu.

It is obviously true that the correct instance name needs to be used when running a command that contains it. I think it also provides some clarity in the text if the instances are referenced in this way, although it might warrant an explanatory note to highlight the distinction up-front.

CarlosNihelton replied:
Remember we have a reference item about this naming convention in our docs:

https://canonical-ubuntu-wsl.readthedocs-hosted.com/en/latest/reference/distributions/#naming

edibotopic replied:
Indeed.

edibotopic commented:
@CarlosNihelton --- I updated the PR with some of the following changes:

- Used "Ubuntu (Preview)", as discussed.
- Removed `#` as a prompt symbol. We do not (currently) use it in these docs and I think we should avoid using a third symbol if possible (if needed we can just specify `root`). In addition, a side-effect of including them now is that it creates problems in Python code blocks, where `#` followed by whitespace can be interpreted as a prompt.
- Add prompt symbols throughout the docs because I think if we're going to use them then we need to do it consistently.
- Added a section to `contributing.md` on how the code blocks are implemented and things that a contributor should be aware of.

CarlosNihelton reviewed (approved):
Thanks for the fix!
